contracts/paymaster/src/paymaster.rs

```diff
@@ -4,38 +4,57 @@ use multiversx_sc::imports::*;
 
 pub mod forward_call;
 pub mod paymaster_proxy;
-const FEE_PAYMENT: usize = 0;
+const FEE_PAYMENT_INDEX: usize = 0;
 
 #[multiversx_sc::contract]
 pub trait PaymasterContract: forward_call::ForwardCall {
     #[init]
     fn init(&self) {}
 
+    #[upgrade]
+    fn upgrade(&self) {}
+
     #[endpoint(forwardExecution)]
     #[payable("*")]
     fn forward_execution(
         &self,
         relayer_addr: ManagedAddress,
         dest: ManagedAddress,
+        min_gas_limit: u64,
         endpoint_name: ManagedBuffer,
         endpoint_args: MultiValueEncoded<ManagedBuffer>,
     ) {
+        let original_caller = self.blockchain().get_caller();
```

Comment:
Not mandatory, but these lines can be moved in a different verification function, to keep this one more clear.